### DISCUSSION TITLE
Integration with rust-tools.nvim

### DISCUSSION THREAD

**Discussion by mightyiam** [VonHeikemen/lsp-zero.nvim#5] [Q&A] [answered]
How would I go about integrating this with the following plugin, please?
https://github.com/simrat39/rust-tools.nvim

**Comment by VonHeikemen** [comment]
It seems `rust-tools` calls the language server inside their setup function. Means you have to use that. Here is an example.

```lua
local lsp = require('lsp-zero')
lsp.preset('system-lsp')

local custom_attach = function(client, bufnr)
  -- some code
end

lsp.on_attach(custom_attach)

lsp.setup_servers({'html', 'cssls'})
lsp.setup()

-- rust-tools setup

local lsp_capabilities = require('cmp_nvim_lsp').update_capabilities(
  vim.lsp.protocol.make_client_capabilities()
)

require('rust-tools').setup({
  server = {
    capabilities = lsp_capabilities,
    on_attach = function(client, bufnr)
      custom_attach(client, bufnr)
      -- more code
    end,
  }
})
```

Edit: sadly you'll have to setup your own keybindings in `on_attach` function for rust.

**Reply by mightyiam** [reply]
What additional API would rust-tools need to have for perfect integration?

**Reply by VonHeikemen** [reply]
It would be lovely if `rust-tools` would let you control the call to the server.

```lua
require('rust-tools').setup({
  server = function()
    lsp.configure('rust_analyzer')
  end
})
``` 

But for the people using the `recommended` preset it would not work. I would have to add another option to setup the server. Something like this.

```lua
lsp.configure('rust_analyzer', {
  setup_client = function(opts, call)
    require('rust-tools').setup({server = opts})
  end
})
```

**Reply by VonHeikemen** [reply]
I came up with another idea: add a function that can generate the options necessary to call the language server using `nvim-lspconfig`.

The updated example for `system-lsp` preset would be this.

```lua
local lsp = require('lsp-zero')
lsp.preset('system-lsp')

lsp.setup_servers({'html', 'cssls'})

lsp.setup()

-- Initialize rust_analyzer with rust-tools
local lsp_rust = lsp.build_options('rust_analyzer', {})
require('rust-tools').setup({server = lsp_rust})
```

For the `recommended` preset `.build_options` should be called before `lsp.setup()`.

```lua
local lsp = require('lsp-zero')
lsp.preset('recommended')

local lsp_rust = lsp.build_options('rust_analyzer', {})

lsp.setup()

-- Initialize rust_analyzer with rust-tools
require('rust-tools').setup({server = lsp_rust})
```

When you use `.build_options` the language server you specify will be not be setup in `lsp-zero`. This way there is no conflict with plugins like `rust-tools`.

**Comment by mightyiam** [comment, marked as answer]
Since 3b6387e242c9441f5ad008aff150c8b882fd86f3 [there is a way](https://github.com/VonHeikemen/lsp-zero.nvim#build_optionsserver-opts)!
